src/XmlScope.ts

```diff
@@ -153,6 +153,46 @@ export class XmlScope extends Scope {
         });
     }
 
+    public getCodeActions(file: BscFile, range: Range) {
+        const result = [] as CodeAction[];
+        for (const diagnostic of this.diagnostics) {
+            //skip diagnostics that don't occur on this line
+            if (diagnostic.range?.start.line !== range.start.line) {
+                continue;
+            }
+            if (diagnostic.code === DiagnosticCodeMap.callToUnknownFunction) {
+                //functionName is stored on this specific diagnostic
+                const lowerFunctionName = (diagnostic as any).functionName.toLowerCase();
```

Comment:
I feel like these will get quite large pretty quickly, so it might be useful to extract it into some type of codeAction-specific helper file. Thoughts?

Reply:
I definitely don't approve the pattern of adding code-actions logic inside Scope/BrsFile/XmlFile. If anything we need less logic (including validation) in those files.

Logic should be entirely decoupled from these classes, and possibly even moved into an internal plugin.

Reply:
Done. moved into `bscPlugin/BscPlugin.ts`

Reply:
I know you 2 have a different opinion; but I'm of the opposite school of thought. I think Roku app are small and this language will only ever be for Roku apps so my opinion is augment all the things early and thoroughly so as to decorate the way with as much info as possible to facilitate its purpose, which surely is transpiling be and supporting the ide. 

I believe Bron has some clever ideas about how to do this that won't incur the tax of supporting a shadow-ast or the cost (perf and conceptionally) of synchronizing the ide specific metadata with the ast. 

But I'm still staunchly of the opposite opinion.  What do we want? Ide features and diagnostics? When do we want them? Now!! :) Ie get as much as we can as fast as we can. 

Sharing this for your benefit @elsassph as Bron knows my views on this already. I'm very solid in the idea that we should never forget what we are building. These apps will never ever be more than a few mb big. That's my ocams razor right there. You 2 will ignore me, of course; but at least you know where I'm coming from.

Reply:
Not sure what you are referring to @georgejecook. Is it about my suggestion to move this logic into an internal plugin?

Reply:
>I definitely don't approve the pattern of adding code-actions logic inside Scope/BrsFile/XmlFile

Reply:
I personally say, if it's clean and manageable, put it wherever gives us the most info, without the need for more parses, or complex data structure sync.

Reply:
Using an internal plugin won't require more data or ast duplication and stuff.

Reply:
On the other hand the amount of code in BrsFile is getting our of control.